force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,66 +154,142 @@ public with sharing class DMLOperations {
+                    AccountId = accProvided.Id,
+                    StageName = 'Prospecting',  // Set a default stage
+                    CloseDate = Date.today().addDays(30)  // Set a default close date
+                );
+            }
+            oppsToUpsert.add(opp);
+        }
+    
+        // Perform the upsert operation
+        upsert oppsToUpsert;
+    }
     /**
      * Question 9
-	 * Upserts an Account with the specified name.
-	 * If an Account with the given name already exists, update the description to "Updated Account".
-	 * If no such Account exists, it creates a new one and sets the description to "New Account".
-	 * @param accountName The name of the Account to be upserted.
-	 * @return The Account that was upserted.
-	 */
-	public static Account upsertAccount(String accountName) {
+     * Upserts an Account with the specified name.
+     * If an Account with the given name already exists, update the description to "Updated Account".
+     * If no such Account exists, it creates a new one and sets the description to "New Account".
+     * @param accountName The name of the Account to be upserted.
+     * @return The Account that was upserted.
+     */
+    public static Account upsertAccount(String accountName) {
 		// Query for the Account
-		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
-
-		// if the Account exists, update the description
-
-		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
+		List<Account> accounts = [SELECT Id, Name, Description FROM Account WHERE Name = :accountName];
+        Account acc; //declare Account object
+        if(accounts.isEmpty()){
+            acc = new Account(Name = accountName, Description = 'New Account');
+        } //if the Account list is empty, initialize the Account object
+        else {
+            acc = accounts[0];
+            acc.Description = 'Updated Account';
+        } //if not, set the account to the one found with the query and update the description
+        upsert acc; 
+		return acc; 
 	}
-
     /**
      * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+     * Manage the association between Contacts and Accounts
+     * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+     * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
```

Comment:
You could have followed the hint and shortened up your code quite a bit in the following method:
`(Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)`